src/persist-client/src/internal/state.rs

```diff
@@ -64,6 +66,51 @@ pub struct HollowBatch<T> {
     pub keys: Vec<PartialBatchKey>,
     /// The number of updates in the batch.
     pub len: usize,
+    /// Runs of sequential batch parts, stored as indices into [keys].
+    ///
+    /// ex. keys=[k1, k2, k3], runs=[]     --> all keys are part of a single run
+    ///     keys=[k1, k2, k3], runs=[1]    --> runs are [k1] and [k2, k3]
+    ///     keys=[k1, k2, k3], runs=[1, 2] --> runs are [k1], [k2], [k3]
+    pub runs: Vec<usize>,
```

Comment:
hm wasn't sure about this. `Vec<Vec<_>>` is certainly more ergonomic, but in the same way you were suspicious of a large # of diffs in a single incremental state change, I'm a little wary of a ton of runs in a single batch, e.g. from a initial snapshot or other large not-necessarily-ordered `append` calls. And keeping track of just indices allows to use slices everywhere and avoid allocs for vecs and key cloning. My hope was the iterator helps abstract over some of the complications with dealing with the raw indices

Reply:
the ergonomics certainly take a hit if HollowBatch also uses this representation. one thing that makes me less worried about this is that I don't ever see us decreasing the blob target size below 8MiB or so, so the number of these we could possibly create is pretty decently limited

my inclination here is to do the efficient thing for serialization (harder to change, insulated entirely within the RustType impl) and then start here with the obvious and ergonomic thing. then we let the perf issues be led by profiling. however, I'm also happy for us to do some BOE math and jump right to the latter if I'm being silly

Reply:
If we keep the serialization format + do what I think is involved to fix https://github.com/MaterializeInc/materialize/pull/14527#discussion_r960853740, we'll still keep a bit of the dance around how to go back and forth between the representations, so it might just move the complexity... are there specific spots that were tricky to read when reviewing? The other thing I was pondering was compaction seems like the only user of runs, so something like a `computed` reader should never have to worry about them. Anyhoo I'll play around with it today and see how it goes

For BOE math, I'm not sure if this is realistic, but if we had 100GiB worth of different runs, we'd be doing ~25K extra allocations each time we pull down state. I don't have a good sense whether that's one drop in the midst of a vast ocean, or a meaningful increase over what state takes right now.

Reply:
> so it might just move the complexity
my starting premise here is exactly that it's better to move the complexity entirely into the RustType impl

anyway, sounds like this one was half baked, too. ditto feel free to ignore until I get a chance to really load the details up in my head and develop a more informed opinion

Reply:
thought about this more. this is easy to change later (no migration), so willing to defer to your judgement here

Reply:
I'll leave it as-is for now, but I'm open to revisiting in the future! hoping to avoid letting this PR linger into the offsite week